File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/data/database/RideRealmDataStore.java
package com.lattis.ellipse.data.database;

import com.lattis.ellipse.data.database.base.RealmObservable;
import com.lattis.ellipse.data.database.mapper.RealmRideMapper;
import com.lattis.ellipse.data.database.model.RealmRide;
import com.lattis.ellipse.domain.model.Ride;

import javax.inject.Inject;

import io.reactivex.Observable;
import io.reactivex.functions.Function;
import io.realm.Realm;
import io.realm.RealmConfiguration;

/**
 * Created by ssd3 on 4/4/17.
 */

public class RideRealmDataStore {

    private RealmConfiguration realmConfiguration;
    private RealmRideMapper rideMapper;

    @Inject
    public RideRealmDataStore(RealmConfiguration realmConfiguration,
                              RealmRideMapper rideMapper) {
        this.realmConfiguration = realmConfiguration;
        this.rideMapper = rideMapper;
    }

    public Observable<Ride> getRide() {
        return RealmObservable.object(
                realm -> {
                    RealmRide realmRide = realm.where(RealmRide.class)
                        .findFirst();
                    return realmRide!=null ? realmRide : new RealmRide();

                })
                .map(realmRide -> rideMapper.mapOut(realmRide));
    }

    public Observable<Ride> createOrUpdateUser(Ride ride) {
        return RealmObservable.object(
                new Function<Realm, RealmRide>() {
                    @Override
                    public RealmRide apply(Realm realm) {
                        return realm.copyToRealmOrUpdate(rideMapper.mapIn(ride));
                    }
                })
                .map(new Function<RealmRide, Ride>() {
                    @Override
                    public Ride apply(RealmRide realmRide) {
                        return rideMapper.mapOut(realmRide);
                    }
                });
    }


    public Observable<Boolean> deleteRide() {
        return RealmObservable.deleteObject(
                realm -> {
                    RealmRide realmRide = realm.where(RealmRide.class)
                            .findFirst();
                    if(realmRide != null){
                        realmRide.deleteFromRealm();
                        return true;
                    }
                    return false;
                });
    }
}

File: velolabs_repos/lattis_android_v2/data/src/main/java/com/lattis/data/database/store/RideRealmDataStore.kt
package com.lattis.data.database.store

import com.lattis.data.database.base.RealmObservable
import com.lattis.data.database.mapper.RealmRideMapper
import com.lattis.data.database.model.RealmRide
import com.lattis.domain.models.Ride
import javax.inject.Inject

import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.functions.Function
import io.realm.Realm
import io.realm.RealmConfiguration


/**
 * Created by ssd3 on 4/4/17.
 */

class RideRealmDataStore @Inject
constructor(private val realmConfiguration: RealmConfiguration,
            private val rideMapper: RealmRideMapper
) {

    fun getRide(): Observable<Ride> {
        return RealmObservable.`object`<RealmRide>(
            Function { realm ->
                val realmRide = realm.where(RealmRide::class.java)
                    .findFirst()
                if (realmRide != null) realmRide else RealmRide()

            })
            .map { realmRide -> rideMapper.mapOut(realmRide) }
    }

    fun createOrUpdateUser(ride: Ride): Observable<Ride> {
        return RealmObservable.`object`(
                Function<Realm, RealmRide> { realm -> realm.copyToRealmOrUpdate(rideMapper.mapIn(ride)) })
                .map { realmRide -> rideMapper.mapOut(realmRide) }
    }


    fun deleteRide(): Observable<Boolean> {
        return RealmObservable.deleteObject(
                Function{ realm ->
                    val realmRide = realm.where(RealmRide::class.java)
                            .findFirst()
                    if (realmRide != null) {
                        realmRide!!.deleteFromRealm()
                        true
                    }
                    false
                })
    }
}
